### PR TITLE
Implementing activation behavior for <label>

### LINKED ISSUE #8179: Implement 'activation behavior' for <label> elements
https://html.spec.whatwg.org/multipage/#the-label-element:the-label-element-10

### PR DESCRIPTION
**Pull request by martiansideofthemoon**
Attempt to resolve #8179 
@Manishearth , could you give me some resources having more information about what each function in `Activatable` does? The code compiles on my machine but I guess a lot more is needed

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8260)

### CODE REVIEW

File: components/script/dom/htmllabelelement.rs [hunk 1]
```diff
@@ -40,6 +43,39 @@ impl HTMLLabelElement {
...
+    // https://html.spec.whatwg.org/multipage/#run-post-click-activation-steps
+    fn activation_behavior(&self, _event: &Event, _target: &EventTarget) {
+        self.upcast::<Element>()
+            .as_maybe_activatable()
+            .map(|a| a.synthetic_click_activation(false, false, false, false));
+    }
+
+    // https://html.spec.whatwg.org/multipage/#implicit-submission
+    fn implicit_submission(&self, ctrlKey: bool, shiftKey: bool, altKey: bool, metaKey: bool) {
+
```
Comment (Manishearth) [inline components/script/dom/htmllabelelement.rs]: Add a FIXME here and file an issue for investigation

**Comment by Manishearth** [comment]
Does this change any tests?

**Comment by Manishearth** [comment]
> could you give me some resources having more information about what each function in Activatable does? 

See the spec links :smile: 

Basically, form controls, buttons, and links can be "activated". This happens when they get clicked, or sometimes when you hit the space/enter key when they are in focus. There's a lot of complications in the spec around this, and the `Activatable` trait abstracts it away. Anything implementing this just needs to fill in the compulsory methods, and it's good to go. In the case of label there was no mention of canceled/post-activation steps (unlike, say, checkbox input), so it's empty.

File: components/script/dom/htmllabelelement.rs (hunk 1 above)
Comment (martiansideofthemoon) [inline components/script/dom/htmllabelelement.rs]: What would be the appropriate FIXME message / issue title?

Comment (Manishearth) [inline components/script/dom/htmllabelelement.rs]: File one, "Investigate and implement implicit submission for label elements"

Comment (martiansideofthemoon) [inline components/script/dom/htmllabelelement.rs]: #8263

**Comment by martiansideofthemoon** [comment]
@Manishearth `./mach test-ref`, `./mach test-unit`passed all tests. Which test has failed? How could I see the same? Didn't find it on pressing the "Details" link

**Comment by Manishearth** [comment]
No, there weren't test failures, that was an induced failure (I was testing some new infrastructure). I was talking about the wpt tests.

**Comment by martiansideofthemoon** [comment]
What would be the correct way of checking this? I am currently running `mach test-wpt` and it looks like it's going to take a really long time to complete.

**Comment by Manishearth** [comment]
You can pass `--processes 4` or something to it.

(Also, use the `--log-raw` parameter as described in tests/wpt/README.md so you don't need to rerun it to update expectations)

**Comment by eefriedman** [comment]
You probably need to change https://github.com/servo/servo/blob/285e29c06637f31a8b8a27c2e454468717924ebd/components/script/dom/element.rs#L1746 before this will actually do anything.

**Comment by Manishearth** [comment]
Good point, forgot to check for that.

**Comment by martiansideofthemoon** [comment]
@Manishearth What change will be needed in that file?

**Comment by Manishearth** [comment]
Add a case to HTMLLabelElement to that match

File: components/script/dom/htmllabelelement.rs [hunk 2]
```diff
@@ -40,6 +43,39 @@ impl HTMLLabelElement {
...
+    // https://html.spec.whatwg.org/multipage/#run-post-click-activation-steps
+    fn activation_behavior(&self, _event: &Event, _target: &EventTarget) {
+        self.upcast::<Element>()
+            .as_maybe_activatable()
+            .map(|a| a.synthetic_click_activation(false, false, false, false));
+    }
+
+    // https://html.spec.whatwg.org/multipage/#implicit-submission
+    fn implicit_submission(&self, ctrlKey: bool, shiftKey: bool, altKey: bool, metaKey: bool) {
+        //FIXME: Investigate and implement implicit submission for label elements
```
Comment (Manishearth) [inline components/script/dom/htmllabelelement.rs]: Could you file a bug for this and refer to the bug number here?

**Comment by Manishearth** [comment]
@bors-servo try

Let's see if there are test issues

**Comment by Manishearth** [comment]
Small nit, then r=me unless there are test failures

File: components/script/dom/htmllabelelement.rs (hunk 2 above)
Comment (martiansideofthemoon) [inline components/script/dom/htmllabelelement.rs]: #8263

Comment (Manishearth) [inline components/script/dom/htmllabelelement.rs]: I mean update the fixme comment here to point to that

**Comment by Ms2ger** [comment]
```
./components/script/dom/element.rs:1726: trailing whitespace
```

**Pull request merged**